### DISCUSSION TITLE
Extract polygon holes in a polygon

### DISCUSSION THREAD

**Discussion by RasmusGrift96** [mozman/ezdxf#397] [Q&A]
Hi,

I'm currently working on a project where I want to make a standalone program to convert images to .dxf files. I want to use these .dxf files in solid works as sketches.
The problem I'm having is if there are holes in my polygon. How can I get the .dxf to detect it is a hole instead of a new polygon.
Take for example the next image:
![teststar'](https://user-images.githubusercontent.com/81967363/114872044-34e8f800-9dfa-11eb-9b18-3fd42c23fb09.png)

If I run my code, I detect the contours with their respective hierarchy (child, parent etc.) with the [cv2.findContours()](https://opencv-python-tutroals.readthedocs.io/en/latest/py_tutorials/py_imgproc/py_contours/py_contours_begin/py_contours_begin.html) function. So I have that information. The next step is that i'm connecting the dots with the layout.add_line() function which results in 2 polygons (circle and a star). If I convert it to .dxf. It doesn't know the circle is actually the parent of the star and it uploads as 2 seperate polygons as below.
![out](https://user-images.githubusercontent.com/81967363/114873630-c311ae00-9dfb-11eb-9ba4-54cd33b2c7d4.png)

This is my code:
```
import cv2
import numpy as np

from typing import Tuple, List
import ezdxf
from ezdxf.document import Drawing
from ezdxf.addons.drawing.matplotlib import qsave


Point = Tuple[float, float]


def create_dxf_from_lines(lines: List[Tuple[Point, Point]]) -> Drawing:
    doc = ezdxf.new()
    layout = doc.modelspace()
    
    for a, b in lines:
        layout.add_line(a, b)

    return doc


def main():

    img = cv2.imread('teststar.png', cv2.IMREAD_UNCHANGED)
    img = cv2.bitwise_not(img)

    #convert img to grey
    img_grey = cv2.cvtColor(img,cv2.COLOR_BGR2GRAY)
    #set a thresh
    thresh = 100
    #get threshold image
    ret,thresh_img = cv2.threshold(img_grey, thresh, 255, cv2.THRESH_BINARY)
    #find contours
    contours, hierarchy = cv2.findContours(thresh_img, cv2.RETR_TREE, cv2.CHAIN_APPROX_SIMPLE)
    
    lines = []
    #connect the coordinates of the contours.
    for contour in range(len(contours)):
        for point in range (len(contours[contour])):
            if point < len(contours[contour])-1:
                Point1=(contours[contour][point][0][0], contours[contour][point][0][1])
                Point2= (contours[contour][point+1][0][0], contours[contour][point+1][0][1])
                lines.append((Point1,Point2))
            else:
                #connect the last 2 coordinates
                lines.append(((contours[contour][point][0][0], contours[contour][point][0][1]), (contours[contour][0][0][0], contours[contour][0][0][1])))

    doc = create_dxf_from_lines(lines)
    doc.saveas('out.dxf')
    qsave(doc.modelspace(), 'out.png')
    
    #create an empty image for contours
    img_contours = np.zeros(img.shape)
    # draw the contours on the empty image
    cv2.drawContours(img_contours, contours, 0, (0,255,0), 1)
    #save image
    cv2.imwrite('contours.png',img_contours) 

    
if __name__ == '__main__':
    main()
```

I've read something about surface, region, polyface etc in the documentation but I'm not really sure which one I should use. 

Thanks in advance
R

**Comment by mbway** [comment]
it's not very intuitive, but i think HATCH is the entity you are looking for. HATCH entities can have solid color in addition to patterns and support holes.
For example:
https://github.com/mozman/ezdxf/issues/225#issuecomment-702236209
